### DISCUSSION TITLE
Should Genplugin generate a tests directory and placeholder unit test?

### DISCUSSION THREAD

**Discussion by Coffee2Bits** [DevOps-MBSE/AaC#92] [Archive] [answered]
Should Genplugin generate a tests directory and placeholder unit test?

I think it would be helpful to our plugin developers if we generate/scaffold the project as much as possible, including generating a placeholder unit test.

**Comment by cvchaparro** [comment]
I think this would be a great addition because it also shows that we encourage testing plugins and even require tests for contributed plugins.

**Comment by jondavid-black** [comment, marked as answer]
I agree it should create a tests directory.

As I'm trying to implement the last bits of the spec plugin I keep running into issues.  I look at the last 2 plugins you guys write and see they have a different directory structure.  I'm also not getting the hookspec generated for the extension.

Here's my follow-on questions:
1) should plugin source go into a directory inside the folder containing yaml like you guys have done?
2) why is the extension hookspec not being generated by "aac gen-plugin ..."?

**Reply by jondavid-black** [reply]
I've been digging around a bit in the genplug.py and think I've found the issue...but it may just be some python mechanic that I don't know or understand.  On line 146 there is a list defined in a way I don't understand.  It looks like this may be preventing the "non-model" types from being processed.

In general, code inside a list declaration is strange and complex to me.  I'd rather not use such things even if it is valid python.  As a python noob I can tell you it is intimidating.

**Reply by jondavid-black** [reply]
Well, that theory was wrong.  I still don't understand those lines of python but the real issue was a mismatch between the Jinja2 template and the properties being provided.  When I change the key in genplug.py line 154 to extensions it generated the hookspec for the extensions.

**Reply by jondavid-black** [reply]
All that said, we should write a couple more tests for gen-plug to keep this from recurring in the future.

**Reply by Coffee2Bits** [reply]
I've logged the bug and solution here: https://github.com/jondavid-black/AaC/issues/100

**Comment by lizzcondrey** [comment]
This has been marked resolved, has an answer, and has the bug and merge linked. Closing this discussion.
